### DISCUSSION TITLE
Maintenance mode does not move volumes of offline VMs

### DISCUSSION THREAD

**Discussion by Rubueno** [apache/cloudstack#9886] [Ideas]
Currently when enabling maintenance mode on a host, VMs are offloaded. What about volumes that are present on the local storage pool of VMs that are powered off? In case of for example wanting to delete the host, the operator must offload these volumes manually. Is it an idea to include this in maintenance mode?

**Comment by rajujith** [comment]
@Rubueno  I believe the volumes in local storage won't be migrated according to the current design. However, having a tool or script to evacuate a primary storage will be useful.

**Reply by Rubueno** [reply]
Is it an idea to build this into the maintenance mode?

**Reply by rajujith** [reply]
@Rubueno  you may create a feature request ticket or you could implement it yourself if possible.

**Reply by DaanHoogland** [reply]
@Rubueno , this not seem to belong with maintenance mode persé. A host in maintenance mode can be reenabled and you wouldn't want to have crippled the VMs by deleting their disks. I would say it would be part of the delete itself. Does that make sense? cc @rajujith

**Reply by Rubueno** [reply]
Maybe we should implement a new feature entirely then. We do not want to delete the disks, but we do want disks of powered off VMs offloaded to another host so that we can do with the host as we see fit. We would like the enable maintenance mode and offloading active and stopped VMs to be fully hands off in all cases.

**Reply by DaanHoogland** [reply]
@rajujith @Rubueno, so the feature is: "As an operator I want an API to move all disks from the local storage of a host in maintenance to other hosts"?
Can you distil this further and add an issue for it? This is no guarantee anybody will implement it but it does make sense to have it on the backlog of feature requests.

**Reply by rajujith** [reply]
@DaanHoogland what you proposed is a good option. I am adding more details, this will be an admin tool/API to migrate volumes in bulk /batch.
- Optionally select the target primary storage. 
- This operation doesn't have to be tied to host maintenance. 
- without creating chaos where cloudstack / operator can control the batch size etc. A queueing mechanism to limit parallel migrations. 
- If possible, migrate from local to shared primary storage as well.

**Reply by DaanHoogland** [reply]
sounds good @rajujith , how about from primary to secondary as well?

**Reply by rajujith** [reply]
> sounds good @rajujith , how about from primary to secondary as well?

I wouldn't migrate to secondary.

**Reply by Rubueno** [reply]
Thanks for the suggestion @rajujith. Sounds good.

**Comment by rajujith** [comment]
@Rubueno  I just learned there is this setting host.maintenance.local.storage.strategy from https://github.com/apache/cloudstack/issues/9887

**Reply by Rubueno** [reply]
Yes, that determines what will happen to running local storage VMs when enabling maintenance mode. @BartJM is in my company. We're working on several bugfixes.
